### PR TITLE
#414 - JDK 17 Upgrade Support Series - Filestore & Pipeline model config

### PR DESCRIPTION
**Pull request by habibimoiz**
#414 - JDK 17 Upgrade Support Series - Filestore & Pipeline model config
- Modifications to ensure FileStore and Pipeline model config is functional with the JDK 17 upgrade.
- Fixing an outdated aiSSEMBLE docs link in a FileStore template file.

### CODE REVIEW

File: foundation/foundation-core-java/src/main/java/com/boozallen/aissemble/core/filestore/AbstractFileStore.java
Comment (habibimoiz) [inline foundation/foundation-core-java/src/main/java/com/boozallen/aissemble/core/filestore/AbstractFileStore.java:1]: Repairs NPE error that was occurring when the `_FS_OVERRIDES` property was not provided.

File: foundation/foundation-core-java/src/main/java/com/boozallen/aissemble/core/filestore/EnvironmentVariableFileStoreConfig.java
Comment (habibimoiz) [inline foundation/foundation-core-java/src/main/java/com/boozallen/aissemble/core/filestore/EnvironmentVariableFileStoreConfig.java:1]: Repairs NPE error that was occurring when the `_FS_OVERRIDES` property was not provided.

File: foundation/foundation-mda/src/main/resources/templates/file-store/file.store.impl.java.vm
Comment (habibimoiz) [inline foundation/foundation-mda/src/main/resources/templates/file-store/file.store.impl.java.vm:1]: Fixing broken link to the pipeline-metalmodel doc.

**Comment by habibimoiz** [comment]
OTS completed with @carter-cundiff & @cwoods-cpointe.

File: foundation/foundation-core-java/src/main/java/com/boozallen/aissemble/core/filestore/AbstractFileStore.java [hunk 1]
```diff
@@ -45,10 +45,14 @@ public AbstractFileStore(final String name) {
      * @return a fully connected context
      */
     public BlobStoreContext openContext() {
-        return ContextBuilder.newBuilder(fileStoreConfig.getProvider())
-                .credentials(fileStoreConfig.getAccessKeyId(), fileStoreConfig.getSecretAccessKey())
-                .overrides(fileStoreConfig.getOverrides())
-                .buildView(BlobStoreContext.class);
+        ContextBuilder contextBuilder = ContextBuilder.newBuilder(fileStoreConfig.getProvider());
+        contextBuilder.credentials(fileStoreConfig.getAccessKeyId(), fileStoreConfig.getSecretAccessKey());
```
Comment (carter-cundiff) [inline foundation/foundation-core-java/src/main/java/com/boozallen/aissemble/core/filestore/AbstractFileStore.java:49]:
```suggestion
        ContextBuilder contextBuilder = ContextBuilder.newBuilder(fileStoreConfig.getProvider())
            .credentials(fileStoreConfig.getAccessKeyId(), fileStoreConfig.getSecretAccessKey());
```

Comment (carter-cundiff) [inline foundation/foundation-core-java/src/main/java/com/boozallen/aissemble/core/filestore/AbstractFileStore.java:49]: Can be updated to chain these together:
```
        ContextBuilder contextBuilder = ContextBuilder.newBuilder(fileStoreConfig.getProvider())
            .credentials(fileStoreConfig.getAccessKeyId(), fileStoreConfig.getSecretAccessKey());
```

Comment (habibimoiz) [reply foundation/foundation-core-java/src/main/java/com/boozallen/aissemble/core/filestore/AbstractFileStore.java]: Chained them together.

File: foundation/foundation-core-java/src/main/java/com/boozallen/aissemble/core/filestore/AbstractFileStore.java [hunk 2]
```diff
@@ -45,10 +45,14 @@ public AbstractFileStore(final String name) {
...
      */
     public BlobStoreContext openContext() {
-        return ContextBuilder.newBuilder(fileStoreConfig.getProvider())
-                .credentials(fileStoreConfig.getAccessKeyId(), fileStoreConfig.getSecretAccessKey())
-                .overrides(fileStoreConfig.getOverrides())
-                .buildView(BlobStoreContext.class);
+        ContextBuilder contextBuilder = ContextBuilder.newBuilder(fileStoreConfig.getProvider())
+                .credentials(fileStoreConfig.getAccessKeyId(), fileStoreConfig.getSecretAccessKey());
+
+        if (!fileStoreConfig.getOverrides().isEmpty()) {
```
Comment (ewilkins-csi) [inline foundation/foundation-core-java/src/main/java/com/boozallen/aissemble/core/filestore/AbstractFileStore.java:51]: Q: Why is this change necessary? Does an empty overrides map cause issues? I'd think since the getter never returns `null` then it'd be safe to just pass it in.

Comment (habibimoiz) [reply foundation/foundation-core-java/src/main/java/com/boozallen/aissemble/core/filestore/AbstractFileStore.java:51]: Yes, when the Overrides map is empty, we are getting a NPE and it can't invoke the map. To avoid the NPE from being thrown, the logic will now check to see if that map is null, if its not then it will chain it to the `ContextBuilder`.

File: foundation/foundation-core-java/src/main/java/com/boozallen/aissemble/core/filestore/EnvironmentVariableFileStoreConfig.java [hunk 3]
```diff
@@ -58,11 +59,12 @@ public String getSecretAccessKey() {
...
-        final Gson gson = new Gson();
-        final Map overrides = gson.fromJson(overridesJsonString, Map.class);
-
         final Properties propertyOverrides = new Properties();
-        propertyOverrides.putAll(overrides);
+        if (overridesJsonString != null) {
+            final Gson gson = new Gson();
+            final Map overrides = gson.fromJson(overridesJsonString, Map.class);
+            propertyOverrides.putAll(overrides);
+        }
```
Comment (JeffreyRoss) [inline foundation/foundation-core-java/src/main/java/com/boozallen/aissemble/core/filestore/EnvironmentVariableFileStoreConfig.java:67]: Q: Why the need for the null check?

File: foundation/foundation-core-java/src/main/java/com/boozallen/aissemble/core/filestore/AbstractFileStore.java (hunk 2 above)
Comment (carter-cundiff) [reply foundation/foundation-core-java/src/main/java/com/boozallen/aissemble/core/filestore/AbstractFileStore.java:51]: @habibimoiz I don't think we actually tested with the empty map, we only tested with a null map. It might be fine to always add the map now that we guarantee `.getOverrides()` will never return null.

Comment (ewilkins-csi) [reply foundation/foundation-core-java/src/main/java/com/boozallen/aissemble/core/filestore/AbstractFileStore.java:51]: This actually ends up causing a new NPE if the contract for `getOverrides` changes and `null` is sometimes returned. FWIW, `ContextBuilder` would throw a NPE if `getOverrides` did return `null`. So we could null-guard this instead of calling `isEmpty`. Or we could just clarify in the JavaDocs of `getOverrides` that it returns an empty map if the value hasn't been provided.

Comment (habibimoiz) [reply foundation/foundation-core-java/src/main/java/com/boozallen/aissemble/core/filestore/AbstractFileStore.java:51]: I am in favor of null-guarding it. 

Proposed change: 
```
    public BlobStoreContext openContext() {
        ContextBuilder contextBuilder = ContextBuilder.newBuilder(fileStoreConfig.getProvider())
                .credentials(fileStoreConfig.getAccessKeyId(), fileStoreConfig.getSecretAccessKey());

        if (fileStoreConfig.getOverrides() != null) {
            contextBuilder.overrides(fileStoreConfig.getOverrides());
        }
        return contextBuilder.buildView(BlobStoreContext.class);
    }
```

Comment (carter-cundiff) [reply foundation/foundation-core-java/src/main/java/com/boozallen/aissemble/core/filestore/AbstractFileStore.java:51]: Let's verify is still works with an empty overrides as well.

**Review by ewilkins-csi — APPROVED** [review]
Looks good barring test of empty overrides map

File: foundation/foundation-core-java/src/main/java/com/boozallen/aissemble/core/filestore/AbstractFileStore.java (hunk 2 above)
Comment (habibimoiz) [reply foundation/foundation-core-java/src/main/java/com/boozallen/aissemble/core/filestore/AbstractFileStore.java]: Just tested out with both null and empty overrides and I am getting the expected outcome!

**Comment by habibimoiz** [comment]
> Looks good barring test of empty overrides map

Just tested both empty and null and its not throwing an NPE error.
